### ISSUE TITLE
Crash on lazy variables inside of closures

### ISSUE THREAD

**Issue by rnapier** [apple/swift#59329]
**Describe the bug**
Starting in Swift 5.7 (Xcode 14b1), creating a lazy variable inside a closure can crash the compiler.

**To Reproduce**

```
enum E { case a }

class X{ init(e: E) { fatalError() } }
class Y{}

let a = {
    lazy var b = X(e: .a)
    return Y()
}()
```

**Expected behavior**
This code should compile. It crashes as follows:

```
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  lldb                     0x0000000102b3d60c llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  lldb                     0x0000000102b3cc00 llvm::sys::RunSignalHandlers() + 112
2  lldb                     0x0000000102b3dcb8 SignalHandler(int) + 304
3  libsystem_platform.dylib 0x00000001a722f4a4 _sigtramp + 56
4  LLDB                     0x000000011381be5c swift::ParameterListInfo::ParameterListInfo(llvm::ArrayRef<swift::AnyFunctionType::Param>, swift::ValueDecl const*, bool) + 216
5  LLDB                     0x0000000113233a98 matchCallArguments(swift::constraints::ConstraintSystem&, swift::FunctionType*, swift::ArgumentList*, llvm::ArrayRef<swift::AnyFunctionType::Param>, llvm::ArrayRef<swift::AnyFunctionType::Param>, swift::constraints::ConstraintKind, swift::constraints::ConstraintLocatorBuilder, llvm::Optional<swift::constraints::TrailingClosureMatching>, llvm::SmallVectorImpl<std::__1::pair<swift::TypeVariableType*, swift::OpenedArchetypeType*> >&) + 384
6  LLDB                     0x0000000113232aec swift::constraints::ConstraintSystem::simplifyApplicableFnConstraint(swift::Type, swift::Type, llvm::Optional<swift::constraints::TrailingClosureMatching>, swift::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder) + 2540
7  LLDB                     0x000000011323991c swift::constraints::ConstraintSystem::addConstraintImpl(swift::constraints::ConstraintKind, swift::Type, swift::Type, swift::constraints::ConstraintLocatorBuilder, bool) + 936
8  LLDB                     0x000000011321e03c swift::constraints::ConstraintSystem::addConstraint(swift::constraints::ConstraintKind, swift::Type, swift::Type, swift::constraints::ConstraintLocatorBuilder, bool) + 76
9  LLDB                     0x00000001132063c4 (anonymous namespace)::ConstraintGenerator::visitApplyExpr(swift::ApplyExpr*) + 656
10 LLDB                     0x00000001131fd270 swift::ASTVisitor<(anonymous namespace)::ConstraintGenerator, swift::Type, void, void, void, void, void>::visit(swift::Expr*) + 160
11 LLDB                     0x00000001131fb178 (anonymous namespace)::ConstraintWalker::walkToExprPost(swift::Expr*) + 628
12 LLDB                     0x00000001135e7a08 (anonymous namespace)::Traversal::doIt(swift::ArgumentList*) + 168
13 LLDB                     0x00000001135e7bf8 swift::ASTVisitor<(anonymous namespace)::Traversal, swift::Expr*, swift::Stmt*, bool, swift::Pattern*, bool, void>::visit(swift::Expr*) + 404
14 LLDB                     0x00000001135e7350 swift::Expr::walk(swift::ASTWalker&) + 80
15 LLDB                     0x00000001131f7b60 swift::constraints::ConstraintSystem::generateConstraints(swift::Expr*, swift::DeclContext*, bool) + 152
16 LLDB                     0x00000001131f70b4 swift::constraints::ConstraintSystem::generateConstraints(swift::constraints::SolutionApplicationTarget&, swift::FreeTypeVariableBinding) + 984
17 LLDB                     0x00000001131f60f8 (anonymous namespace)::ClosureConstraintGenerator::visitDecl(swift::Decl*) + 580
18 LLDB                     0x00000001131f2644 swift::constraints::ConstraintSystem::simplifyClosureBodyElementConstraint(swift::ASTNode, swift::constraints::ContextualTypeInfo, bool, swift::OptionSet<swift::constraints::ConstraintSystem::TypeMatchFlags, unsigned int>, swift::constraints::ConstraintLocatorBuilder) + 656
19 LLDB                     0x000000011323ac8c swift::constraints::ConstraintSystem::simplifyConstraint(swift::constraints::Constraint const&) + 1692
20 LLDB                     0x000000011324b45c swift::constraints::ConjunctionElement::attempt(swift::constraints::ConstraintSystem&) const + 232
21 LLDB                     0x00000001132567f8 swift::constraints::ConjunctionStep::attempt(swift::constraints::ConjunctionElement const&) + 240
22 LLDB                     0x0000000113257808 swift::constraints::BindingStep<swift::constraints::ConjunctionElementProducer>::take(bool) + 504
23 LLDB                     0x00000001132469b4 swift::constraints::ConstraintSystem::solveImpl(llvm::SmallVectorImpl<swift::constraints::Solution>&) + 400
24 LLDB                     0x000000011324873c swift::constraints::ConstraintSystem::solve(llvm::SmallVectorImpl<swift::constraints::Solution>&, swift::FreeTypeVariableBinding) + 60
25 LLDB                     0x0000000113248394 swift::constraints::ConstraintSystem::solveImpl(swift::constraints::SolutionApplicationTarget&, swift::FreeTypeVariableBinding) + 504
26 LLDB                     0x0000000113247d94 swift::constraints::ConstraintSystem::solve(swift::constraints::SolutionApplicationTarget&, swift::FreeTypeVariableBinding) + 148
27 LLDB                     0x000000011336e604 swift::TypeChecker::typeCheckExpression(swift::constraints::SolutionApplicationTarget&, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 496
28 LLDB                     0x000000011336f41c swift::TypeChecker::typeCheckBinding(swift::Pattern*&, swift::Expr*&, swift::DeclContext*, swift::Type, swift::PatternBindingDecl*, unsigned int, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 196
29 LLDB                     0x000000011336f614 swift::TypeChecker::typeCheckPatternBinding(swift::PatternBindingDecl*, unsigned int, swift::Type, swift::OptionSet<swift::TypeCheckExprFlags, unsigned int>) + 292
30 LLDB                     0x000000011341c0e0 swift::PatternBindingEntryRequest::evaluate(swift::Evaluator&, swift::PatternBindingDecl*, unsigned int) const + 2000
31 LLDB                     0x00000001133ae190 llvm::Expected<swift::PatternBindingEntryRequest::OutputType> swift::Evaluator::getResultUncached<swift::PatternBindingEntryRequest>(swift::PatternBindingEntryRequest const&) + 268
32 LLDB                     0x00000001133ae024 swift::PatternBindingEntryRequest::OutputType swift::evaluateOrDefault<swift::PatternBindingEntryRequest>(swift::Evaluator&, swift::PatternBindingEntryRequest, swift::PatternBindingEntryRequest::OutputType) + 108
33 LLDB                     0x00000001133a3f28 swift::ASTVisitor<(anonymous namespace)::DeclChecker, void, void, void, void, void, void>::visit(swift::Decl*) + 3332
34 LLDB                     0x00000001133a0f9c (anonymous namespace)::DeclChecker::visit(swift::Decl*) + 192
35 LLDB                     0x00000001133a0ec4 swift::TypeChecker::typeCheckDecl(swift::Decl*, bool) + 120
36 LLDB                     0x0000000113418528 swift::ASTVisitor<(anonymous namespace)::StmtChecker, void, swift::Stmt*, void, void, void, void>::visit(swift::Stmt*) + 3252
37 LLDB                     0x0000000113416c84 bool (anonymous namespace)::StmtChecker::typeCheckStmt<swift::BraceStmt>(swift::BraceStmt*&) + 136
38 LLDB                     0x0000000113416d00 swift::TypeChecker::typeCheckTopLevelCodeDecl(swift::TopLevelCodeDecl*) + 52
39 LLDB                     0x0000000113441e2c swift::TypeCheckSourceFileRequest::evaluate(swift::Evaluator&, swift::SourceFile*) const + 208
40 LLDB                     0x00000001134435a0 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultUncached<swift::TypeCheckSourceFileRequest>(swift::TypeCheckSourceFileRequest const&) + 268
41 LLDB                     0x0000000113443404 llvm::Expected<swift::TypeCheckSourceFileRequest::OutputType> swift::Evaluator::getResultCached<swift::TypeCheckSourceFileRequest, (void*)0>(swift::TypeCheckSourceFileRequest const&) + 68
42 LLDB                     0x0000000113441cf0 swift::performTypeChecking(swift::SourceFile&) + 48
43 LLDB                     0x00000001123eb4bc lldb_private::SwiftExpressionParser::Parse(lldb_private::DiagnosticManager&, unsigned int, unsigned int) + 316
44 LLDB                     0x00000001123fa528 lldb_private::SwiftUserExpression::Parse(lldb_private::DiagnosticManager&, lldb_private::ExecutionContext&, lldb_private::ExecutionPolicy, bool, bool) + 1372
45 LLDB                     0x000000011223b03c lldb_private::UserExpression::Evaluate(lldb_private::ExecutionContext&, lldb_private::EvaluateExpressionOptions const&, llvm::StringRef, llvm::StringRef, std::__1::shared_ptr<lldb_private::ValueObject>&, lldb_private::Status&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*, lldb_private::ValueObject*) + 1988
46 LLDB                     0x0000000112239184 lldb_private::REPL::IOHandlerInputComplete(lldb_private::IOHandler&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) + 2348
47 LLDB                     0x00000001121a901c lldb_private::IOHandlerEditline::Run() + 284
48 LLDB                     0x000000011218f22c lldb_private::Debugger::RunIOHandlers() + 140
49 LLDB                     0x00000001121929b8 lldb_private::Debugger::IOHandlerThread(void*) + 24
50 LLDB                     0x0000000112244d6c lldb_private::HostNativeThreadBase::ThreadCreateTrampoline(void*) + 120
51 libsystem_pthread.dylib  0x00000001a721826c _pthread_start + 148
52 libsystem_pthread.dylib  0x00000001a721308c thread_start + 8
```

**Environment (please complete the following information):**
 - OS: macOS 12.4
 - Xcode Version/Tag/Branch: 14.0 beta 1

swiftlang-5.7.0.113.202 clang-1400.0.16.2

See also FB10077315

**Comment by xedin** [comment]
This is actually a duplicate of https://github.com/apple/swift/issues/59294 which is fixed by https://github.com/apple/swift/pull/59301. It should be available in the next main and 5.7 toolchains you can download from [swift.org/download](https://www.swift.org/download/) to verify and close this.

**Comment by xedin** [comment]
Changes made it into the snapshot so I'm going to close this out, please feel free to re-open if necessary.
